tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/FunctionGraph.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright 2020 The TensorFlow Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.tensorflow;
+
+import java.util.List;
+import java.util.ListIterator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.math.Sign;
+import org.tensorflow.proto.framework.SignatureDef;
+import org.tensorflow.proto.framework.TensorInfo;
+
+/**
+ * A graph that can be invoked as a single function, with an input and output signature.
+ *
+ * <p>A function can also invoke a
+ * <a href="https://www.tensorflow.org/api_docs/python/tf/function">tf.function</a>
+ * defined in a {@link SavedModelBundle}.
+ *
+ * <pre>{@code
+ * FunctionGraph myFunction = savedModelBundle.function("myFunctionSignatureName");
+ * Map<String, Tensor<?>> outputTensorMap = myFunction.call(inputTensorMap);
+ * }</pre>
+ */
+public class FunctionGraph implements AutoCloseable {
```

Comment:
Maybe `FunctionSession` as it's a specialised session? That way it conceptually lives next to `EagerSession` and  `Session`, rather than next to `Graph`. I feel like this is much closer to a session than a graph.

Reply:
Right now, it is true that all functions loaded from a saved model share the same graph and are just used to execute it with a given signature (therefore they are acting more as a session). Nonetheless, I still have a preference to preserve the `FunctionGraph` naming. 

If we ignore the saving part (which is limited due to the actual state of the C API), conceptually these functions also allow you to build your graphs, replacing the need of allocating explicitly a `Graph` instance, e.g.
```java
public class MyModel {

  // Function graph #1
  private static Signature addTwo(Ops tf) {
    Placeholder<TFloat32> input = tf.placeholder(TFloat32.DTYPE);
    Add<TFloat32> output = tf.math.add(input, tf.constant(2.0f));
    return Signature.builder("addTwo").input("x", input).output("y", output).build();
  }

  // Function graph #2
  private static Signature minusTwo(Ops tf) {
    Placeholder<TFloat32> input = tf.placeholder(TFloat32.DTYPE);
    Add<TFloat32> output = tf.math.sub(input, tf.constant(2.0f));
    return Signature.builder("subTwo").input("x", input).output("y", output).build();
  }

  public static void main(String args[]) {
    try (FunctionGraph f1 = FunctionGraph.create(MyModel::addTwo);
         FunctionGraph f2 = FunctionGraph.create(MyModel::minusTwo)) {
        ... // execute functions (saving both of them in the same model won't work for now)
    }
  }
}
```
Each function has its own graph and therefore if very coupled with that concept (where `EagerSession` has no graph at all). They will also appear as separate graphs when exporting them into a saved model the same way Python does it (in fact, Python saves each function as one or more "objects", which are then linked to their distinct graph in the function library).

Reply:
Another idea would be to call this class a `ConcreteFunction`. Concrete functions and functions are two very similar but distinct concepts in [TF Python](https://www.tensorflow.org/guide/intro_to_graphs). The former is a typed realization of a function graph while the latter is its polymorphic version, only acting as a facade (e.g. a function is eventually composed of multiple concrete functions, one for each type of operands the function has been called with).

In our scenario here, the function graphs are strongly typed and refer to a single graph, therefore more behave like a concrete function in the Python paradigm and it would make sense to name them after it.

We can probably support polymorphic functions too in the future, where we pass the type of the input tensors in parameter to the function builder. So we will still need to find a proper name for this concept, which will probably result in class encapsulating one or more `ConcreteFunction`. Maybe simply `PolymorphicFunction`?

Reply:
My preference for a non-graph name is that `Graph`s aren't executable, and are missing necessary state because the `Variable`s live in the `Session`. So `FunctionGraph` is more like a `Session` because it contains all the necessary bits to execute. However I'm also fine with the `ConcreteFunction` name as that's just a new concept that people have to learn.